### PR TITLE
Added Reset Simulation

### PR DESCRIPTION
**Pull request by triccyx**
Added to clock gazebo plugin the possibility to reset simulation via rpc call

### CODE REVIEW

**Comment by traversaro** [comment]
I guess you want to merge the modification in the `devel` branch, right? 
You can change the target branch of the PR directly from the GitHub web user interface.

**Comment by traversaro** [comment]
Modifying this thrift file is unfortunately extremely tricky, because we have multiple copies of this file disseminated across several repository, see for example https://github.com/robotology/WB-Toolbox/commit/af1fa4106a4597254b28b4f3eaecb41261600124 . I guess @diegoferigo and @S-Dafarra are the maintainers of the repositories affected by this. 

Before merging this modification, we should make sure that the client generated with the old version of the thrift file are still compatible with the server exposed by the new version of the thrift. 

The clean solution is probably to avoid this duplication. This involves either:
* Add to `gazebo-yarp-plugins` all the CMake machinery necessary for making sure that the `thrift/clock/clock_rpc.thrift` file is installed and can be found by the system. The problem of this solution would be that all the repository that use that file will now depend on gazebo-yarp-plugins and thus will depend indirectly on Gazebo. This is definitely problematic, think for example on how this will affect Windows.
* Add this interface to YARP: if we are able to ensure that this interface is relatively simulator-independent, it would make sense to have this in YARP. On a positive note, at the moment the interface is quite independent from the underlying simulator, as long as it is a fixed time-step simulator as the "Step" functions take the number of steps as input, so it could make sense to have this in YARP.

**Comment by diegoferigo** [comment]
@triccyx I remember a discussion with @francesco-romano (previous WB-Toolbox maintainer) about removing that thrift-related sources as soon as YARP would have installed them in the system. I'd go as well in this direction rather than storing it into `gazebo-yarp-plugins`.

**Comment by diegoferigo** [comment]
Cross ref https://github.com/robotology/wb-toolbox/issues/62

**Comment by triccyx** [comment]
What is the best way to resolve the problem? Without this merge the tests won't work

**Comment by diegoferigo** [comment]
I'd like to see this thrift to end into yarp. Without spending too much time to figure out if it is or isn't good enough, since it seems already sufficiently generic, why not opening a PR and ask for its inclusion? It might be a quick process. @traversaro

**Comment by traversaro** [comment]
I agree with @diegoferigo .

**Comment by drdanz** [comment]
I think we need to discuss this... At the moment I wouldn't know 
* where to put the thrift interfaces
* whether we want to include the thrift only, or to generate a library as well
* In case we want to add a library, we need to decide if we want it static or shared
* if shared we need to modify the thrift generator to allow to handle the dll import/export
* Also we need to commit the generated files in the repository

I suggest to merge this PR and to have a meeting about this after YARP 3 release

**Comment by triccyx** [comment]
Ok for Add ClockServer thrift file #1724, but this doesn't solve the problem. 
The client at the moment can't execute the tests.  Perhaps we can merge these modifications in a branch in master and ask the client to use that branch.

**Comment by traversaro** [comment]
If clients with the old interface are still able to call the methods in the new interface (I think should be the case, is anyone is able to check this?), I don't think it would be too problematic to merge this PR.

However, I think the current method added is a bit misleading, as it is called `reset`, but it actually just reset the state of the simulation, and not the time. I think `resetSimulation` should reset both time and state of the simulation, and eventually a `resetSimulationState`should just reset the state without resetting the time.

**Comment by triccyx** [comment]
Any news/solutions regarding this pull request?

**Comment by traversaro** [comment]
Can you address my comments from the 4th of June? 

> However, I think the current method added is a bit misleading, as it is called reset, but it actually just reset the state of the simulation, and not the time. I think resetSimulation should reset both time and state of the simulation, and eventually a resetSimulationStateshould just reset the state without resetting the time.

**Comment by triccyx** [comment]
Ok I change the method name to ResetSimulation as it resets both state and time.

**Comment by triccyx** [comment]
ping

**Comment by traversaro** [comment]
Ok for me, but I would prefer if @diegoferigo could check this as well.

File: plugins/clock/src/Clock.cc [hunk 1]
```diff
@@ -177,6 +177,12 @@ namespace gazebo
         m_world->ResetTime();
     }
 
+    void GazeboYarpClock::resetSimulation()
+    {
+        //m_world->ResetEntities(gazebo::physics::Base::MODEL);
```
Comment (traversaro) [inline plugins/clock/src/Clock.cc]:
```suggestion
```
I think we can remove this.

**Comment by triccyx** [comment]
@diegoferigo friendly ping

**Comment by traversaro** [comment]
> Can you confirm this?

Exactly. As you can check in the generated code for the server https://github.com/robotology/gazebo-yarp-plugins/pull/345/files#diff-9e60d5f99e3cf3eb4705af443b70c244R395 , nothing in the code changes for methods different from `resetSimulation`. 
The only possible incompatibility is related to some new software that tries to call `resetSimulation` on an old version of `gazebo-yarp-plugins`.

>     Would the signal to reset the world be propagated to all the other plugins? E.g. if there are controllers, their state should be reset too (I think yes, just checking).

Plugins can handle the reset even by attaching a callback to the `WorldReset` event, see how the controlboard plugin handle this:
* https://github.com/robotology/gazebo-yarp-plugins/blob/01ad1d1252b0827033fd57058f066cc790b893c6/plugins/controlboard/src/ControlBoardDriver.cpp#L342
* https://github.com/robotology/gazebo-yarp-plugins/blob/01ad1d1252b0827033fd57058f066cc790b893c6/plugins/controlboard/src/ControlBoardDriver.cpp#L624
old related issue and PR : https://github.com/robotology/gazebo-yarp-plugins/issues/92 and https://github.com/robotology/gazebo-yarp-plugins/pull/247 

> Btw, I am still looking for storing all these thrift files somewhere else, in a centralized place in order to avoid to keep vendoring these files in other projects.

I agree.

**Comment by diegoferigo** [comment]
Ack. I'll keep in mind to update the wb-toolbox thrift in any case :+1:

**Pull request merged**